builder/dockerfile/internals.go

```diff
@@ -125,9 +136,32 @@ func (b *Builder) performCopy(state *dispatchState, inst copyInstruction) error
 		return err
 	}
 
+	chownPair := b.archiver.IDMappings.RootPair()
+	// if a chown was requested, perform the steps to get the uid, gid
+	// translated (if necessary because of user namespaces), and replace
+	// the root pair with the chown pair for copy operations
+	if inst.flChown != "" {
+		userStr, grpStr := parseChownFlag(inst.flChown)
+		uid, err := lookupUser(userStr, filepath.Join(destInfo.root, "etc", "passwd"))
+		if err != nil {
+			return errors.Wrapf(err, "can't find uid for user "+userStr)
+		}
+		gid, err := lookupGroup(grpStr, filepath.Join(destInfo.root, "etc", "group"))
+		if err != nil {
+			return errors.Wrapf(err, "can't find gid for group "+userStr)
+		}
+
+		// convert as necessary because of user namespaces
+		chownPair, err = b.archiver.IDMappings.ToHost(idtools.IDPair{UID: uid, GID: gid})
+		if err != nil {
+			return errors.Wrapf(err, "unable to convert uid/gid to host mapping")
+		}
+	}
```

Comment:
Please extract this whole block (parsing the uid/gid pair) into a new function, and maybe call that function from `dispatchCopy()` instead of here.